### PR TITLE
drivers: sensor: add driver for amg88xx sensor

### PR DESCRIPTION
**Pull request by jfischer-no**
This patch adds the driver for Panasonic AMG88xx infrared array sensor.

The driver was developed within the framework of a student internship in the development department. The task of the student was to implement the trigger part of the driver, transfer and visualization of the measured data.

The sensor needs an array like `struct sensor_value temp_value[64];` to transfer the data. I'm not sure if the driver is allowed to use the sensor API this way.

### CODE REVIEW

File: drivers/sensor/amg88xx/Kconfig [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	help
+	  Enable driver for AMG88XX infrared thermopile sensor.
+
+config AMG88XX_NAME
+	string
+	prompt "Driver name"
+	default "AMG88XX"
+	depends on AMG88XX
+	help
+	  Device name with which the AMG88XX sensor is identified.
```
Comment (dbkinder) [inline drivers/sensor/amg88xx/Kconfig]: How about:

     Device name identifying the AMG88XX sensor.

File: drivers/sensor/amg88xx/Kconfig [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+config AMG88XX_I2C_ADDR
+	hex
+	prompt "I2C address for TMP006 Sensor"
```
Comment (dbkinder) [inline drivers/sensor/amg88xx/Kconfig]: Should that be ``AMG88XX`` instead of ``TMP006``?

File: drivers/sensor/amg88xx/Kconfig [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	  0x68: AD-SELECT connected GND
+	  0x69: AD-SELECT connected VDD
+
+config AMG88XX_I2C_MASTER_DEV_NAME
+	string
+	prompt "I2C master where AMG88XX is connected"
+	depends on AMG88XX
+	default "I2C_0"
+	help
+	  Specify the device name of the I2C master device to which the
```
Comment (dbkinder) [inline drivers/sensor/amg88xx/Kconfig]: How about:

     Specify the I2C master device's name where the AMG88XX sensor is connected.

File: drivers/sensor/amg88xx/Kconfig [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	  AMG88XX chip is connected.
+
+choice
+	prompt "Trigger mode"
+	depends on AMG88XX
+	default AMG88XX_TRIGGER_NONE
+	help
+	  Specify the type of triggering to be used by the driver.
```
Comment (dbkinder) [inline drivers/sensor/amg88xx/Kconfig]: delete "to be"

File: drivers/sensor/amg88xx/Kconfig [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	bool
+	depends on AMG88XX
+
+config AMG88XX_GPIO_DEV_NAME
+	string
+	prompt "GPIO device"
+	default "GPIO_0"
+	depends on AMG88XX && AMG88XX_TRIGGER
+	help
+	  The device name of the GPIO device to which the AMG88XX interrupt
```
Comment (dbkinder) [inline drivers/sensor/amg88xx/Kconfig]: How about:

     The GPIO device's name where the AMG88XX interrupt (alert) pin is connected.

File: drivers/sensor/amg88xx/Kconfig [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	  (alert) pin is connected.
+
+config AMG88XX_GPIO_PIN_NUM
+	int
+	prompt "Interrupt GPIO pin number"
+	default 0
+	depends on AMG88XX && AMG88XX_TRIGGER
+	help
+	  The number of the GPIO pin on which the interrupt signal from the
```
Comment (dbkinder) [inline drivers/sensor/amg88xx/Kconfig]: How about:

     The GPIO pin number receiving the interrupt signal from the AMG88XX sensor.

File: samples/sensor/amg88xx/README.rst [hunk 7]
```diff
@@ -0,0 +1,69 @@
+.. _amg88xx:
+
+AMG88XX Infrared Array Sensor
+#############################
+
+Overview
+********
+
+Sample application that periodically reads 8x8 temperature array from
```
Comment (dbkinder) [inline samples/sensor/amg88xx/README.rst]: How about:

     This sample application periodically reads the 8x8 temperature array from
     the AMG88XX sensor, and displays its values to the serial console.

File: samples/sensor/amg88xx/README.rst [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+the AMG88XX sensor. Sample can also be configured to be triggered
+when the upper or lower threshold has been reached.
+
+Requirements
+************
+
+Panasonic AMG88XX is an 8x8 (64) pixel infrared array sensor.
```
Comment (dbkinder) [inline samples/sensor/amg88xx/README.rst]: The Panasonic ....

File: samples/sensor/amg88xx/README.rst [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+AMG88XX Infrared Array Sensor
+#############################
+
+Overview
+********
+
+Sample application that periodically reads 8x8 temperature array from
+the AMG88XX sensor. Sample can also be configured to be triggered
+when the upper or lower threshold has been reached.
```
Comment (dbkinder) [inline samples/sensor/amg88xx/README.rst]: I'm confused a bit here.  Is the upper/lower threshold an application-supplied value that will cause a trigger when it get's reached?  If so, where does the application specify this value?

File: samples/sensor/amg88xx/README.rst [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+We used the Evalutation Kit mounted on FRDM-K64F board.
+On the Evalutation Kit, all jumpers except the J11 must be removed.
+
+This sample uses the sensor APIs and the provided driver for the AMG88xx sensor.
+
+Building and Running
+********************
+
+This project outputs sensor data to the console. If you want to test
```
Comment (dbkinder) [inline samples/sensor/amg88xx/README.rst]: How about:

     This sample reads from the sensor and outputs sensor data to the console every second.
     If you want to test the sensor's trigger mode, specify the trigger configuration in the Kconfig
     file and wire the interrupt output from the sensor to your board.  For the FRDM-K64F we tested,
     pin PTC2 must be muxed as GPIO.

I'm not sure of the technical details here, but you should clearly state how to configure and hook up the sensor.

File: samples/sensor/amg88xx/README.rst [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+   000|02075 01900 02000 01925 02050 01975 01950 01900
+   001|02000 02000 02000 01900 02125 02000 02050 02050
+   002|02075 01950 01950 01925 02050 01975 02075 02000
+   003|02175 02000 01975 01800 02000 02125 01925 02050
+   004|01975 02050 02050 02000 02000 01975 01925 02000
+   005|02100 02100 02075 02000 02025 02050 02100 02100
+   006|01975 02150 02075 02025 02025 02225 02025 02375
+   007|02025 02075 02150 02075 02025 02050 02125 02400
+
+<repeats endlessly>
```
Comment (dbkinder) [inline samples/sensor/amg88xx/README.rst]: How about:

     The sensor array data is output every second.

**Review by dbkinder — CHANGES_REQUESTED** [review]
some suggested edits for clarity and grammar

File: drivers/sensor/Kconfig [hunk 12]
```diff
@@ -110,4 +110,6 @@ source "drivers/sensor/tmp007/Kconfig"
 
 source "drivers/sensor/tmp112/Kconfig"
 
+source "drivers/sensor/amg88xx/Kconfig"
```
Comment (MaureenHelm) [inline drivers/sensor/Kconfig]: Move this change up so the sensors stay in alpha order

File: drivers/sensor/Makefile [hunk 13]
```diff
@@ -36,3 +36,4 @@ obj-$(CONFIG_SX9500) += sx9500/
 obj-$(CONFIG_TH02) += th02/
 obj-$(CONFIG_TMP007) += tmp007/
 obj-$(CONFIG_TMP112) += tmp112/
+obj-$(CONFIG_AMG88XX) += amg88xx/
```
Comment (MaureenHelm) [inline drivers/sensor/Makefile]: Move this change up so the sensors stay in alpha order

File: drivers/sensor/amg88xx/Kconfig [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+	depends on I2C
+	default n
+	help
+	  Enable driver for AMG88XX infrared thermopile sensor.
+
+config AMG88XX_NAME
+	string
+	prompt "Driver name"
+	default "AMG88XX"
+	depends on AMG88XX
```
Comment (MaureenHelm) [inline drivers/sensor/amg88xx/Kconfig]: Let's wrap this config and everything below it with ``if AMG88XX`` so we don't keep repeating `depends on AMG88XX`

File: drivers/sensor/amg88xx/Kconfig [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+#
+# Copyright (c) 2017 Phytec Messtechnik GmbH
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig AMG88XX
+	bool
+	prompt "AMG88XX Infrared Thermopile Sensor"
```
Comment (MaureenHelm) [inline drivers/sensor/amg88xx/Kconfig]: Squash these two lines into one:
```bool "AMG88XX Infrared Thermopile Sensor"```

Do this throughout

File: drivers/sensor/amg88xx/amg88xx.c [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+#include <gpio.h>
+#include <misc/byteorder.h>
+#include <misc/util.h>
+#include <kernel.h>
+#include <sensor.h>
+#include <misc/__assert.h>
+
+#include "amg88xx.h"
+
+int amg88xx_reg_read(struct amg88xx_data *drv_data, u8_t reg, u8_t *val)
```
Comment (MaureenHelm) [inline drivers/sensor/amg88xx/amg88xx.c]: I think these wrapper functions are overkill. We're also beginning to see stack overflows due to really deep call stacks.

File: drivers/sensor/amg88xx/amg88xx.c [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+
+static int amg88xx_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct amg88xx_data *drv_data = dev->driver_data;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_TEMP);
+
+	if (i2c_burst_read(drv_data->i2c, CONFIG_AMG88XX_I2C_ADDR,
+			   AMG88XX_OUTPUT_BASE,
+			   (u8_t *)drv_data->sample, 128)) {
```
Comment (MaureenHelm) [inline drivers/sensor/amg88xx/amg88xx.c]: Can we replace the magic number with `sizeof(drv_data->sample)`?

File: drivers/sensor/amg88xx/amg88xx.c [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+
+#ifdef CONFIG_AMG88XX_TRIGGER
+	if (amg88xx_init_interrupt(dev) < 0) {
+		SYS_LOG_ERR("Failed to initialize interrupt!");
+		return -EIO;
+	}
+#endif
+
+	dev->driver_api = &amg88xx_driver_api;
```
Comment (MaureenHelm) [inline drivers/sensor/amg88xx/amg88xx.c]: Remove this and use `DEVICE_AND_API_INIT` below

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/amg88xx/Kconfig [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+config AMG88XX_NAME
+	string
+	prompt "Driver name"
+	default "AMG88XX"
+	depends on AMG88XX
+	help
+	  Device name identifying the AMG88XX sensor.
+
+config AMG88XX_I2C_ADDR
```
Comment (bogdan-davidoaia) [inline drivers/sensor/amg88xx/Kconfig]: You can also add a range attribute to the config to force it to be either 0x68 or 0x69.

File: drivers/sensor/amg88xx/amg88xx.c [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+		val[idx].val1 = (((s32_t)drv_data->sample[idx]) *
+				  AMG88XX_TREG_LSB_SCALING) / 100;
+		val[idx].val2 = (((s32_t)drv_data->sample[idx]) *
+				  AMG88XX_TREG_LSB_SCALING) % 100;
+	}
+
+	return 0;
+}
+
+int amg88xx_init_device(struct device *dev)
```
Comment (bogdan-davidoaia) [inline drivers/sensor/amg88xx/amg88xx.c]: This function can be static.

File: drivers/sensor/amg88xx/amg88xx.c [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+		if (amg88xx_reg_write(drv_data, AMG88XX_PCLT,
+				       AMG88XX_PCLT_NORMAL_MODE)) {
+			return -EIO;
+		}
+		k_busy_wait(AMG88XX_WAIT_MODE_CHANGE_US);
+	}
+
+	if (amg88xx_reg_write(drv_data, AMG88XX_RST,
+			       AMG88XX_RST_INITIAL_RST)) {
+		k_busy_wait(AMG88XX_WAIT_INITIAL_RESET_US);
```
Comment (bogdan-davidoaia) [inline drivers/sensor/amg88xx/amg88xx.c]: Why is this here? Shouldn't it be after the if statement (i. e., if the write is successful)?

File: drivers/sensor/amg88xx/amg88xx.h [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+		     enum sensor_channel chan,
+		     enum sensor_attribute attr,
+		     const struct sensor_value *val);
+
+int amg88xx_trigger_set(struct device *dev,
+		     const struct sensor_trigger *trig,
+		     sensor_trigger_handler_t handler);
+
+int amg88xx_init_interrupt(struct device *dev);
+#endif
```
Comment (bogdan-davidoaia) [inline drivers/sensor/amg88xx/amg88xx.h]: Nit: here and the next endif put a /* ... */ comment with CONFIG_AMG88XX_TRIGGER and _SENSOR_AMG88XX.

File: drivers/sensor/amg88xx/amg88xx.c [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	for (size_t idx = 0; idx < len; idx++) {
+		/* fix negative values */
+		if (drv_data->sample[idx] & (1 << 11)) {
+			drv_data->sample[idx] |= 0xF000;
+		}
+		val[idx].val1 = (((s32_t)drv_data->sample[idx]) *
+				  AMG88XX_TREG_LSB_SCALING) / 100;
+		val[idx].val2 = (((s32_t)drv_data->sample[idx]) *
```
Comment (bogdan-davidoaia) [inline drivers/sensor/amg88xx/amg88xx.c]: val2 is expressed in micro-units (i.e., val = val1 + val2 / 10^6).

So in addition to the operations you wrote, you need to multiply by 10^4.

After doing this change, you should also update the sample in accordance.

File: drivers/sensor/amg88xx/amg88xx.c [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+	return 0;
+}
+
+int amg88xx_init_device(struct device *dev)
+{
+	struct amg88xx_data *drv_data = dev->driver_data;
+	u8_t tmp;
+
+	if (amg88xx_reg_read(drv_data, AMG88XX_PCLT, &tmp)) {
+		return -EIO;
```
Comment (bogdan-davidoaia) [inline drivers/sensor/amg88xx/amg88xx.c]: SYS_LOG_ERR/DBG message wouldn't hurt for these errors for easier debugging.

**Review by bogdan-davidoaia — CHANGES_REQUESTED** [review]

File: drivers/sensor/amg88xx/amg88xx.c (hunk 24 above)
Comment (punitvara) [reply drivers/sensor/amg88xx/amg88xx.c]: Use SYS_LOG_ERR to show error as other new drivers are using same.

**Review by dbkinder — APPROVED** [review]
+1 for doc changes

**Review by bogdan-davidoaia — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**